### PR TITLE
Make sort inference a preprocessing pass

### PR DESCRIPTION
**Pull request by ajreynol**
This changes the interface of the SortInference module to be multiple steps, which can then be used by the preprocessing architecture.

### CODE REVIEW

File: src/theory/sort_inference.cpp [hunk 1]
```diff
@@ -114,203 +114,201 @@ void SortInference::reset() {
...
-          Trace("sort-inference") << " ";
-        }
-        Trace("sort-inference") << ") -> ";
-        retTn = retTn[(int)retTn.getNumChildren()-1];
+void SortInference::initialize(const std::vector<Node>& assertions)
+{
+  Trace("sort-inference-proc") << "Calculating sort inference..." << std::endl;
+  // process all assertions
+  std::map<Node, int> visited;
+  for (unsigned i = 0; i < assertions.size(); i++)
```
Comment (HanielB) [inline src/theory/sort_inference.cpp]: cache

File: src/theory/sort_inference.cpp [hunk 2]
```diff
@@ -114,203 +114,201 @@ void SortInference::reset() {
...
+  {
+    Trace("sort-inference-debug") << "Process " << assertions[i] << std::endl;
+    std::map<Node, Node> var_bound;
+    process(assertions[i], var_bound, visited);
+  }
+  Trace("sort-inference-proc") << "...done" << std::endl;
+  for (std::map<Node, int>::iterator it = d_op_return_types.begin();
```
Comment (HanielB) [inline src/theory/sort_inference.cpp]: for-each

File: src/theory/sort_inference.cpp [hunk 3]
```diff
@@ -114,203 +114,201 @@ void SortInference::reset() {
...
-      for( unsigned i=0; i<it->first[0].getNumChildren(); i++ ){
-        recordSubsort( it->first[0][i].getType(), it->second[it->first[0][i]] );
-        printSort( "sort-inference", it->second[it->first[0][i]] );
-        Trace("sort-inference") << std::endl;
-      }
+    recordSubsort(retTn, it->second);
+    printSort("sort-inference", it->second);
+    Trace("sort-inference") << std::endl;
+  }
+  for (std::map<Node, std::map<Node, int> >::iterator it = d_var_types.begin();
```
Comment (HanielB) [inline src/theory/sort_inference.cpp]: ditto

File: src/theory/sort_inference.cpp [hunk 4]
```diff
@@ -114,203 +114,201 @@ void SortInference::reset() {
...
+       it != d_var_types.end();
+       ++it)
+  {
+    Trace("sort-inference")
+        << "Quantified formula : " << it->first << " : " << std::endl;
+    for (unsigned i = 0; i < it->first[0].getNumChildren(); i++)
```
Comment (HanielB) [inline src/theory/sort_inference.cpp]: cache

File: src/theory/sort_inference.cpp [hunk 5]
```diff
@@ -114,203 +114,201 @@ void SortInference::reset() {
...
 
-    // simplify all assertions by introducing new symbols wherever necessary
-    Trace("sort-inference-proc") << "Perform simplification..." << std::endl;
-    std::map<Node, std::map<TypeNode, Node> > visited2;
-    for (unsigned i = 0, size = assertions.size(); i < size; i++)
+void SortInference::getNewAssertions(std::vector<Node>& new_asserts)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  // now, ensure constants are distinct
+  for (std::map<TypeNode, std::map<Node, Node> >::iterator it =
```
Comment (HanielB) [inline src/theory/sort_inference.cpp]: for-each

File: src/theory/sort_inference.cpp [hunk 6]
```diff
@@ -114,203 +114,201 @@ void SortInference::reset() {
...
+           d_const_map.begin();
+       it != d_const_map.end();
+       ++it)
+  {
+    std::vector<Node> consts;
+    for (std::map<Node, Node>::iterator it2 = it->second.begin();
```
Comment (HanielB) [inline src/theory/sort_inference.cpp]: ditto

File: src/theory/sort_inference.cpp [hunk 7]
```diff
@@ -114,203 +114,201 @@ void SortInference::reset() {
...
 
-    // enforce constraints based on monotonicity
-    Trace("sort-inference-proc") << "Enforce monotonicity..." << std::endl;
-    for (std::map<TypeNode, std::vector<int> >::iterator it =
-             d_type_sub_sorts.begin();
-         it != d_type_sub_sorts.end();
-         ++it)
+  // enforce constraints based on monotonicity
+  Trace("sort-inference-proc") << "Enforce monotonicity..." << std::endl;
+  for (std::map<TypeNode, std::vector<int> >::iterator it =
```
Comment (HanielB) [inline src/theory/sort_inference.cpp]: ditto

File: src/preprocessing/passes/sort_infer.cpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    for (const Node& na : newAsserts)
+    {
+      Trace("sort-infer-preprocess")
+          << "*** Preprocess SortInferencePass : new constraint " << na << endl;
+      assertionsToPreprocess->push_back(na);
+    }
+    // indicate correspondence between the functions
+    // TODO (#2308): move this to a better place
+    SmtEngine* smt = smt::currentSmtEngine();
+    for (std::map<Node, Node>::iterator it = model_replace_f.begin();
```
Comment (HanielB) [inline src/preprocessing/passes/sort_infer.cpp]: for-each

**Comment by ajreynol** [comment]
I addressed the review.

I also added a regression and added a one line fix for something that was suboptimal in the module (which I realized with the regression).

**Pull request merged**